File: .claude/skills/agentteams/scripts/mjs/src/analyzer.mjs
// Plan analyzer: five-document bundle writer.
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

const DEFAULT_EXPECTATIONS = [
  'schema-present',
  'tests-present',
  'acceptance-gate-defined',
  'dependencies-declared',
  'examples-present',
];

function loadExpectations(planDir) {
  const cfg = path.join(planDir, '.agentteams', 'expectations.yaml');
  if (!fs.existsSync(cfg)) return DEFAULT_EXPECTATIONS;
  const text = fs.readFileSync(cfg, 'utf8');
  const out = [];
  for (const line of text.split('\n')) {
    const m = line.match(/^\s*-\s+(.+?)\s*$/);
    if (m) out.push(m[1]);
  }
  return out.length ? out : DEFAULT_EXPECTATIONS;
}

function checkExpectation(name, plan) {
  const surfaces = [];
  for (const f of plan.features) {
    let status = '❌';
    const evidence = [];
    const hay = (f.acceptanceCriteria.join('\n') + ' ' + f.stories.flatMap(s => s.acceptanceCriteria).join('\n')).toLowerCase();
    if (name === 'schema-present' && /schema|ast|frontmatter|matrix|json/.test(hay)) status = '✅';
    else if (name === 'tests-present' && /test|parity|gate|verify/.test(hay)) status = '✅';
    else if (name === 'acceptance-gate-defined' && /acceptance|gate|exit 0|exit 7|byte-equal/.test(hay)) status = '✅';
    else if (name === 'dependencies-declared' && f.stories.some(s => s.dependencies.length > 0)) status = '✅';
    else if (name === 'examples-present' && /example|fixture|golden|replay/.test(hay)) status = '✅';
    if (status === '❌' && /partial|stub|placeholder/.test(hay)) status = '🟡';
    evidence.push(`${f.path}:1`);
    surfaces.push({ featureId: f.id, slug: f.slug, status, evidence });
  }
  return surfaces;
}

export function buildFitMatrix(plan, expectations = DEFAULT_EXPECTATIONS) {
  const matrix = {};
  for (const name of expectations) matrix[name] = checkExpectation(name, plan);
  return matrix;
}

export function renderFitMatrix(matrix, plan) {
  const expectations = Object.keys(matrix);
  const features = plan.features;
  let out = '# Fit Matrix\n\n| Expectation | ' + features.map(f => `F${f.id}`).join(' | ') + ' |\n';
  out += '| ---: | ' + features.map(() => ':---:').join(' | ') + ' |\n';
  for (const exp of expectations) {
    const cells = matrix[exp].map(s => s.status);
    out += `| ${exp} | ${cells.join(' | ')} |\n`;
  }
  out += '\n## Evidence\n\n';
  for (const exp of expectations) {
    out += `\n### ${exp}\n\n`;
    for (const surface of matrix[exp]) {
      out += `- F${surface.featureId} (${surface.slug}): ${surface.status} — ${surface.evidence.join(', ')}\n`;
    }
  }
  return out;
}

export function renderAnalysis(plan) {
  let out = '# Analysis\n\n';
  out += `Plan: \`${plan.planDir}\`\n\n`;
  for (const f of plan.features) {
    out += `## F${f.id} — ${f.slug}\n\n`;
    out += `Path: \`${f.path}\`\n\n`;
    out += `Stories: ${f.stories.length} | Acceptance criteria: ${f.acceptanceCriteria.length}\n\n`;
    if (f.acceptanceCriteria.length > 0) {
      out += 'Acceptance criteria:\n';
      for (const ac of f.acceptanceCriteria) out += `- ${ac}\n`;
      out += '\n';
    }
  }
  return out;
}

export function renderStateFlow(plan) {
  let out = '# State Flow Audit\n\n';
  const machines = [];
  const re = /([a-z][a-z_-]*\s*(?:→|->)\s*[a-z][a-z_ →\->]*)/gi;
  for (const f of plan.features) {
    let m;
    re.lastIndex = 0;
    const body = (f.acceptanceCriteria.join('\n') + '\n' + f.stories.flatMap(s => s.acceptanceCriteria).join('\n'));
    while ((m = re.exec(body)) !== null) {
      machines.push({ feature: `F${f.id}`, transition: m[1] });
    }
  }
  if (machines.length === 0) {
    out += '_No state-machine transitions detected in plan acceptance criteria._\n';
  } else {
    for (const m of machines) out += `- ${m.feature}: \`${m.transition}\`\n`;
  }
  return out;
}

export function renderSchemaDeltas(plan) {
  let out = '# Schema Deltas\n\n';
  out += '_Schema additions surfaced by gap analysis. Populated by F03 prioritizer._\n\n';
  out += '## Detected schema surfaces\n\n';
  let any = false;
  for (const f of plan.features) {
    const hay = (f.acceptanceCriteria.join('\n') + ' ' + f.stories.flatMap(s => s.acceptanceCriteria).join(' '));
    const schemaHints = (hay.match(/(schema|interface|type|AST|frontmatter)/gi) || []).length;
    if (schemaHints > 0) {
      out += `- **F${f.id}** (${f.slug}): ${schemaHints} schema mentions in acceptance criteria.\n`;
      any = true;
    }
  }
  if (!any) out += '_No explicit schema deltas detected._\n';
  return out;
}

export function renderGapsScaffold(plan) {
  // F02 scaffolds; F03 fills the punch-list.
  return `# Gaps — P0 / P1 / P2

_Scaffold emitted by F02 (analyzer). Populated by F03 (DAG prioritizer)._

## Punch list (placeholder)

Run \`agentteams dag <plan-dir>\` to populate.
`;
}

export function renderReadme(plan, bundleDir) {
  const featureCount = plan.features.length;
  const storyCount = plan.features.reduce((a, f) => a + f.stories.length, 0);
  const taskCount = plan.features.reduce((a, f) => a + f.stories.reduce((b, s) => b + s.tasks.length, 0), 0);
  const warnings = plan.warnings.length;
  return `# Analysis bundle — \`${plan.planDir}\`

## Headline verdict

This plan declares **${featureCount} features**, **${storyCount} stories**, and **${taskCount} tasks**, with **${warnings} parser warnings**. Fit-matrix and gap punch list below.

## Files in this bundle

- \`README.md\` — this file (entry index + headline).
- \`fit-matrix.md\` — expectation × surface scoreboard.
- \`analysis.md\` — long-form per-feature walk-through.
- \`gaps-p0-p1-p2.md\` — prioritized punch list (populated by \`agentteams dag\`).
- \`schema-deltas.md\` — schema additions surfaced by analysis.
- \`state-flow.md\` — state-machine audit.
`;
}

export function writeBundle(plan, outRoot) {
  const uuid = crypto.randomUUID();
  const dir = path.join(outRoot, '.ai-harness', 'analysis', uuid);
  fs.mkdirSync(dir, { recursive: true });
  const expectations = loadExpectations(plan.planDir);
  const matrix = buildFitMatrix(plan, expectations);

  const files = {
    'README.md': renderReadme(plan, dir),
    'fit-matrix.md': renderFitMatrix(matrix, plan),
    'analysis.md': renderAnalysis(plan),
    'gaps-p0-p1-p2.md': renderGapsScaffold(plan),
    'schema-deltas.md': renderSchemaDeltas(plan),
    'state-flow.md': renderStateFlow(plan),
  };
  for (const [name, body] of Object.entries(files)) {
    fs.writeFileSync(path.join(dir, name), body, 'utf8');
  }
  return { uuid, dir, files: Object.keys(files), matrix };
}

File: .claude/skills/agentteams/scripts/py/src/agentteams/analyzer.py
"""Plan analyzer: five-document bundle writer."""
from __future__ import annotations

import re
import uuid as _uuid
from pathlib import Path
from typing import Any

DEFAULT_EXPECTATIONS = [
    "schema-present",
    "tests-present",
    "acceptance-gate-defined",
    "dependencies-declared",
    "examples-present",
]


def _load_expectations(plan_dir: str) -> list[str]:
    cfg = Path(plan_dir) / ".agentteams" / "expectations.yaml"
    if not cfg.exists():
        return list(DEFAULT_EXPECTATIONS)
    out: list[str] = []
    for line in cfg.read_text(encoding="utf-8").split("\n"):
        m = re.match(r"^\s*-\s+(.+?)\s*$", line)
        if m:
            out.append(m.group(1))
    return out or list(DEFAULT_EXPECTATIONS)


def _check_expectation(name: str, plan: dict[str, Any]) -> list[dict[str, Any]]:
    surfaces: list[dict[str, Any]] = []
    for f in plan["features"]:
        status = "❌"
        hay = (
            "\n".join(f["acceptanceCriteria"])
            + " "
            + " ".join(a for s in f["stories"] for a in s["acceptanceCriteria"])
        ).lower()
        if name == "schema-present" and re.search(r"schema|ast|frontmatter|matrix|json", hay):
            status = "✅"
        elif name == "tests-present" and re.search(r"test|parity|gate|verify", hay):
            status = "✅"
        elif name == "acceptance-gate-defined" and re.search(r"acceptance|gate|exit 0|exit 7|byte-equal", hay):
            status = "✅"
        elif name == "dependencies-declared" and any(s.get("dependencies") for s in f["stories"]):
            status = "✅"
        elif name == "examples-present" and re.search(r"example|fixture|golden|replay", hay):
            status = "✅"
        if status == "❌" and re.search(r"partial|stub|placeholder", hay):
            status = "🟡"
        surfaces.append({"featureId": f["id"], "slug": f["slug"], "status": status, "evidence": [f"{f['path']}:1"]})
    return surfaces


def build_fit_matrix(plan: dict[str, Any], expectations: list[str] | None = None) -> dict[str, list[dict[str, Any]]]:
    expectations = expectations or DEFAULT_EXPECTATIONS
    return {name: _check_expectation(name, plan) for name in expectations}


def render_fit_matrix(matrix: dict[str, list[dict[str, Any]]], plan: dict[str, Any]) -> str:
    expectations = list(matrix.keys())
    features = plan["features"]
    header = "| Expectation | " + " | ".join(f"F{f['id']}" for f in features) + " |"
    sep = "| ---: | " + " | ".join(":---:" for _ in features) + " |"
    rows = [
        f"| {exp} | " + " | ".join(s["status"] for s in matrix[exp]) + " |"
        for exp in expectations
    ]
    out = ["# Fit Matrix", "", header, sep, *rows, "", "## Evidence", ""]
    for exp in expectations:
        out.append(f"\n### {exp}\n")
        for surface in matrix[exp]:
            out.append(f"- F{surface['featureId']} ({surface['slug']}): {surface['status']} — {', '.join(surface['evidence'])}")
    return "\n".join(out) + "\n"


def render_analysis(plan: dict[str, Any]) -> str:
    parts = ["# Analysis", "", f"Plan: `{plan['planDir']}`", ""]
    for f in plan["features"]:
        parts.append(f"## F{f['id']} — {f['slug']}\n")
        parts.append(f"Path: `{f['path']}`\n")
        parts.append(f"Stories: {len(f['stories'])} | Acceptance criteria: {len(f['acceptanceCriteria'])}\n")
        if f["acceptanceCriteria"]:
            parts.append("Acceptance criteria:")
            for ac in f["acceptanceCriteria"]:
                parts.append(f"- {ac}")
            parts.append("")
    return "\n".join(parts) + "\n"


def render_state_flow(plan: dict[str, Any]) -> str:
    out = ["# State Flow Audit", ""]
    pat = re.compile(r"([a-z][a-z_-]*\s*(?:→|->)\s*[a-z][a-z_ →\->]*)", re.IGNORECASE)
    found: list[str] = []
    for f in plan["features"]:
        body = "\n".join(f["acceptanceCriteria"] + [a for s in f["stories"] for a in s["acceptanceCriteria"]])
        for m in pat.finditer(body):
            found.append(f"- F{f['id']}: `{m.group(1)}`")
    if not found:
        out.append("_No state-machine transitions detected in plan acceptance criteria._")
    else:
        out.extend(found)
    return "\n".join(out) + "\n"


def render_schema_deltas(plan: dict[str, Any]) -> str:
    out = [
        "# Schema Deltas",
        "",
        "_Schema additions surfaced by gap analysis. Populated by F03 prioritizer._",
        "",
        "## Detected schema surfaces",
        "",
    ]
    any_found = False
    for f in plan["features"]:
        hay = "\n".join(f["acceptanceCriteria"] + [a for s in f["stories"] for a in s["acceptanceCriteria"]])
        hits = len(re.findall(r"schema|interface|type|AST|frontmatter", hay, re.IGNORECASE))
        if hits:
            out.append(f"- **F{f['id']}** ({f['slug']}): {hits} schema mentions in acceptance criteria.")
            any_found = True
    if not any_found:
        out.append("_No explicit schema deltas detected._")
    return "\n".join(out) + "\n"


def render_gaps_scaffold(plan: dict[str, Any]) -> str:
    return (
        "# Gaps — P0 / P1 / P2\n\n"
        "_Scaffold emitted by F02 (analyzer). Populated by F03 (DAG prioritizer)._\n\n"
        "## Punch list (placeholder)\n\n"
        "Run `agentteams dag <plan-dir>` to populate.\n"
    )


def render_readme(plan: dict[str, Any], bundle_dir: str) -> str:
    fc = len(plan["features"])
    sc = sum(len(f["stories"]) for f in plan["features"])
    tc = sum(len(s["tasks"]) for f in plan["features"] for s in f["stories"])
    wc = len(plan["warnings"])
    return (
        f"# Analysis bundle — `{plan['planDir']}`\n\n"
        "## Headline verdict\n\n"
        f"This plan declares **{fc} features**, **{sc} stories**, and **{tc} tasks**, "
        f"with **{wc} parser warnings**. Fit-matrix and gap punch list below.\n\n"
        "## Files in this bundle\n\n"
        "- `README.md` — this file (entry index + headline).\n"
        "- `fit-matrix.md` — expectation × surface scoreboard.\n"
        "- `analysis.md` — long-form per-feature walk-through.\n"
        "- `gaps-p0-p1-p2.md` — prioritized punch list (populated by `agentteams dag`).\n"
        "- `schema-deltas.md` — schema additions surfaced by analysis.\n"
        "- `state-flow.md` — state-machine audit.\n"
    )


def write_bundle(plan: dict[str, Any], out_root: str) -> dict[str, Any]:
    uid = str(_uuid.uuid4())
    out_dir = Path(out_root) / ".ai-harness" / "analysis" / uid
    out_dir.mkdir(parents=True, exist_ok=True)
    expectations = _load_expectations(plan["planDir"])
    matrix = build_fit_matrix(plan, expectations)
    files = {
        "README.md": render_readme(plan, str(out_dir)),
        "fit-matrix.md": render_fit_matrix(matrix, plan),
        "analysis.md": render_analysis(plan),
        "gaps-p0-p1-p2.md": render_gaps_scaffold(plan),
        "schema-deltas.md": render_schema_deltas(plan),
        "state-flow.md": render_state_flow(plan),
    }
    for name, body in files.items():
        (out_dir / name).write_text(body, encoding="utf-8")
    return {"uuid": uid, "dir": str(out_dir), "files": list(files.keys()), "matrix": matrix}
